File: editorial_ch/chap6-26_30/prob6-26.java
import java.util.*;

class Main {
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int Q = sc.nextInt();
		for (int t = 1; t <= Q; t++) {
			// 輸入 → 建構矩陣 A
			double X = sc.nextDouble();
			double Y = sc.nextDouble();
			double Z = sc.nextDouble();
			int T = sc.nextInt();
			double[][] A = {
				{ 1 - X, Y, 0 },
				{ 0, 1 - Y, Z },
				{ X, 0, 1 - Z }
			};

			// 矩陣乘方的計算 → 輸出答案
			double[][] B = copyMatrix(power(A, T));
			double answerA = B[0][0] + B[0][1] + B[0][2];
			double answerB = B[1][0] + B[1][1] + B[1][2];
			double answerC = B[2][0] + B[2][1] + B[2][2];
			System.out.format("%.12f %.12f %.12f\n", answerA, answerB, answerC);
		}
	}
	static final int MOD = 1000000000;
	static double[][] copyMatrix(double[][] A) {
		// 矩陣的複製（注意如果直接以「M = A;」來複製，則只會複製到參照）
		double[][] M = new double[3][3];
		for (int i = 0; i < 3; i++) {
			for (int j = 0; j < 3; j++) {
				M[i][j] = A[i][j];
			}
		}
		return M;
	}
	static double[][] multiplication(double[][] A, double[][] B) {
		// 回傳 3×3 矩陣 A, B 乘積的函式
		double[][] C = { { 0, 0, 0 }, { 0, 0, 0 }, { 0, 0, 0 } };
		for (int i = 0; i < 3; i++) {
			for (int j = 0; j < 3; j++) {
				for (int k = 0; k < 3; k++) {
					C[i][j] += A[i][k] * B[k][j];
				}
			}
		}
		return C;
	}
	static double[][] power(double[][] A, int n) {
		// 回傳 A 的 n 次方的函式
		double[][] P = copyMatrix(A);
		double[][] Q = { { 0, 0, 0 }, { 0, 0, 0 }, { 0, 0, 0 } };
		boolean flag = false;
		for (int i = 0; i < 30; i++) {
			if ((n & (1 << i)) != 0L) {
				if (flag == false) {
					Q = copyMatrix(P);
					flag = true;
				}
				else {
					Q = copyMatrix(multiplication(Q, P));
				}
			}
			P = copyMatrix(multiplication(P, P));
		}
		return Q;
	}
}

File: editorial_ch/chap6-26_30/prob6-26.c
#include <stdio.h>
#include <stdbool.h>

typedef struct {
	double x[3][3];
} matrix;

// 回傳 3×3 的 0 矩陣 (= [[0, 0, 0], [0, 0, 0], [0, 0, 0]])
matrix zero_matrix() {
	matrix A;
	A.x[0][0] = 0; A.x[0][1] = 0; A.x[0][2] = 0;
	A.x[1][0] = 0; A.x[1][1] = 0; A.x[1][2] = 0;
	A.x[2][0] = 0; A.x[2][1] = 0; A.x[2][2] = 0;
	return A;
}

// 回傳 3×3 矩陣乘積的函式
matrix multiplication(matrix A, matrix B) {
	matrix C = zero_matrix();
	int i, j, k;
	for (i = 0; i < 3; i++) {
		for (j = 0; j < 3; j++) {
			for (k = 0; k < 3; k++) {
				C.x[i][j] += A.x[i][k] * B.x[k][j];
			}
		}
	}
	return C;
}

// 回傳矩陣 A 的 n 次方的函式
matrix power(matrix A, int n) {
	matrix P = A;
	matrix Q = zero_matrix();
	bool flag = false;
	for (int i = 0; i < 30; i++) {
		if ((n & (1 << i)) != 0LL) {
			if (flag == false) {
				Q = P;
				flag = true;
			}
			else {
				Q = multiplication(Q, P);
			}
		}
		P = multiplication(P, P);
	}
	return Q;
}

int main() {
	int Q, T; double X, Y, Z;
	scanf("%d", &Q);
	for (int t = 1; t <= Q; t++) {
		// 輸入 → 建構矩陣 A
		scanf("%lf%lf%lf%d", &X, &Y, &Z, &T);
		matrix A = zero_matrix();
		A.x[0][0] = 1.0 - X; A.x[2][0] = X;
		A.x[1][1] = 1.0 - Y; A.x[0][1] = Y;
		A.x[2][2] = 1.0 - Z; A.x[1][2] = Z;
 
		// 矩陣乘方的計算 → 輸出答案
		matrix B = power(A, T);
		double answerA = B.x[0][0] + B.x[0][1] + B.x[0][2];
		double answerB = B.x[1][0] + B.x[1][1] + B.x[1][2];
		double answerC = B.x[2][0] + B.x[2][1] + B.x[2][2];
		printf("%.12lf %.12lf %.12lf\n", answerA, answerB, answerC);
	}
	return 0;
}